fix: handle trailing / in Path.External.drop_prefix

emillon opened the pull request:
In a nutshell:

    # drop_prefix ~prefix:"/a/b/c/" "/a/b/c/d/e"
    Some "d/e"

This improves the test situation on nix (#8203) in the default devshell, because the paths in `OCAMLPATH` can have trailing slashes.

More precisely, the `--sanitize-for-tests` options uses `Path.External.drop_prefix` to replace elements of `OCAMLPATH` by the `FINDLIB` string. If `OCAMLPATH` contains paths with trailing slashes, these prefixes are not sanitized properly.

emillon committed:
fix: handle trailing / in Path.External.drop_prefix

In a nutshell:

    # drop_prefix ~prefix:"/a/b/c/" "/a/b/c/d/e"
    Some "d/e"

This improves the test situation on nix (#8203) in the default devshell,
because the paths in `OCAMLPATH` can have trailing slashes.

More precisely, the `--sanitize-for-tests` options uses
`Path.External.drop_prefix` to replace elements of `OCAMLPATH` by the
`FINDLIB` string. If `OCAMLPATH` contains paths with trailing slashes,
these prefixes are not sanitized properly.

Signed-off-by: Etienne Millon <me@emillon.org>

rgrinberg commented:
The tests are failing.

Also, this needs a better description on why it's better to modify this function rather than the tests themselves.

emillon commented:
I changed to a portable function and amended the description.

rgrinberg commented on otherlibs/stdune/src/path.ml:
```diff
@@ -505,7 +505,7 @@ end = struct
     then Some Local.root
     else (
-      let prefix = prefix ^ "/" in
+      let prefix = if String.is_suffix ~suffix:"/" prefix then prefix else prefix ^ "/" in
```
Can you make sure the `drop_prefix` works in a consistent way for non external paths as well? I think it might rely on the fact that `Path.Local.t` never contains a leading `/`, but it's good to double check.

emillon replied:
I checked and this was failing, so I pushed a fix for this as well

emillon committed:
add test for nonexternal

Signed-off-by: Etienne Millon <me@emillon.org>

emillon committed:
fix nonexternal case

Signed-off-by: Etienne Millon <me@emillon.org>

rgrinberg committed:
refactor: get rid of [Path.External.drop_prefix]

Just as easily replaced by [Path.drop_prefix]. None of its uses were
performance critical, so the reduced complexity is more important here.

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

rgrinberg reviewed (approved):
LGTM. The implementations of drop_prefix were basically copy-pasted so I got rid of the one in `Path.External`. The performance wasn't worth the duplication.

emillon commented:
thanks!
